src/middlewared/middlewared/plugins/cluster_linux/ctdb_private.py

```diff
@@ -0,0 +1,138 @@
+
+        # make sure something was given to us
+        if not data['private_ips']:
+            verrors.add(
+                f'{schema_name}.no_ips',
+                'Private IP(s) must be specified.',
+            )
+
+        verrors.check()
+
+        # normalize the data and remove netmask info (if its given to us)
+        # from `ip` keys since it's not used
+        for i in data['private_ips']:
+            i['ip'] = i['ip'].split('/')[0]
+
+        if schema_name == 'node_create':
+            # get the current ips in the cluster
+            cur_data = {'private_ips': [], 'public_ips': []}
+            cur_data['private_ips'] = await self.query()
+            cur_data['public_ips'] = await self.middleware.call('datastore.query', 'ctdb.public.ips')
+
+            # need to make sure the new private ip(s) don't already exist in the cluster
+            if m := list(set(i['ip'] for i in data['private_ips']).intersection(set(i['ip'] for i in cur_data['private_ips']))):
+                verrors.add(
+                    f'{schema_name}.private_ip_exists',
+                    f'Private IP address(es): {m} already in the cluster.',
+                )
+
+            # need to make sure the new private ip(s) don't already exist in the cluster as public ip(s)
+            if m := list(set(i['ip'] for i in data['private_ips']).intersection(set(i['ip'] for i in cur_data['public_ips']))):
+                verrors.add(
+                    f'{schema_name}.public_ip_exists',
+                    f'Private IP address(es): {m} have already been added to the cluster as public IP(s).',
+                )
+
+        verrors.check()
+
+        return data
+
+    @private
+    async def write_private_ips_to_ctdb(self):
+
+        data = {'private_ips': []}
+        data['private_ips'] = await self.query()
+
+        with open(PRIVATE_IP_FILE, 'w') as f:
+            for i in map(lambda i: i['ip'], data['private_ips']):
+                f.write(f'{i}\n')
+
+        return data
+
+    @accepts(Dict(
+        'node_create',
+        List('private_ips', unique=True, items=[
+            Dict('private_ip', IPAddr('ip'))
+        ], default=[]),
+    ))
+    @job(lock=JOB_LOCK)
+    async def do_create(self, job, data):
+
+        """
+        Create private IPs to be used in the ctdb cluster.
+
+        `private_ips` is a list of `ip` addresses that will be used
+            for the intra-cluster communication between nodes.
+            These ip addresses should be on a phyically separate
+            network and should be non-routable.
+        """
+
+        schema_name = 'node_create'
+        verrors = ValidationErrors()
+
+        data = await self.common_validation(data, schema_name, verrors)
+
+        for i in data['private_ips']:
+            await self.middleware.call(
+                'datastore.insert', 'ctdb.private.ips', i,
+            )
+
+        return await self.write_private_ips_to_ctdb()
```

Comment:
This is a weird CRUD service, since we create a batch of entries? And update a batch of entries.

I don't think it should be a CRUD service at all, or have it be a CRUD method and add extra methods for batch operation...